gh-5436: document best effort ordering of instance dicts (#5443)
pypy/pypy: pypy/doc/cpython_differences.rst

.. __: https://github.com/pypy/pypy/issues/1974

> Order of dictionary keys in instance dicts
> --------------------------------------------

Reviewer: ```suggestion
Order of dictionary keys in instance dicts
------------------------------------------
```

Instance dictionaries are not always ordered by insertion in PyPy.

> Specifically, this can happens if the ``__init__`` of the class adds attributes

Reviewer: ```suggestion
Specifically, this can happen if the ``__init__`` of the class adds attributes
```

in different orders. Here's an example::

>     >>>> class A:

Reviewer: This unfortunately doesn’t get syntax highlighting, but hopefully it will soon with https://github.com/pygments/pygments/pull/3094

Author: cool!

    ....     def __init__(self, a, b, reorder=False):
    ....         if not reorder:
    ....             self.a = a
    ....             self.b = b
    ....         else:
    ....             self.b = b
    ....             self.a = a
    ....
    >>>> print(A(1, 2).__dict__)
    {'a': 1, 'b': 2}
    >>>> print(A(1, 2, reorder=True).__dict__)
    {'a': 1, 'b': 2}

CPython prints ``{'b': 2, 'a': 1}`` in the last line.

Background: After CPython 3.6, dictionaries became `ordered by insertion

> order`__. This is also true for instance dictionaries. In PyPy, regular

Reviewer: ```suggestion
Background: In CPython 3.6, dictionaries became `ordered by insertion`__.
This is also true for instance dictionaries. In PyPy, regular
```
I suggest a little simplification.

dictionaries have been ordered since 2015__. For instance dictionaries, this is
not always possible, because they are represented using `Self-style maps/hidden

> classes`__, which is one of the cornerstones of PyPy's performance.

Reviewer: ```suggestion
classes`__, which are one of the cornerstones of PyPy's performance.
```

.. __: https://github.com/pypy/pypy/issues/1974

Order of dictionary keys in instance dicts
------------------------------------------

Instance dictionaries are not always ordered by insertion in PyPy.
Specifically, this can happen if the ``__init__`` of the class adds attributes
in different orders. Here's an example::

    >>>> class A:
    ....     def __init__(self, a, b, reorder=False):
    ....         if not reorder:
    ....             self.a = a
    ....             self.b = b
    ....         else:
    ....             self.b = b
    ....             self.a = a
    ....
    >>>> print(A(1, 2).__dict__)
    {'a': 1, 'b': 2}
    >>>> print(A(1, 2, reorder=True).__dict__)
    {'a': 1, 'b': 2}

CPython prints ``{'b': 2, 'a': 1}`` in the last line.

> Background: After CPython 3.6, dictionaries became `ordered by insertion

Reviewer: ```suggestion
```
This line is duplicated now, sorry!
